### PR TITLE
Fixing connection error

### PR DESCRIPTION
**Pull request by zezo357**
### New Pull Request Checklist

- [x] I have read the [Documentation](https://pub.dev/documentation/dio/latest/)
- [x] I have searched for a similar pull request in the [project](https://github.com/cfug/dio/pulls) and found none
- [x] I have updated this branch with the latest `main` branch to avoid conflicts (via merge from master or rebase)
- [x] I have added the required tests to prove the fix/feature I'm adding
- [x] I have updated the documentation (if necessary)
- [x] I have run the tests without failures
- [x] I have updated the `CHANGELOG.md` in the corresponding package

### Additional context and info (if any)

fixing https://github.com/cfug/dio/issues/1911

### CODE REVIEW

File: dio/test/basic_test.dart
Comment (kuhnroyal) [inline dio/test/basic_test.dart:47]: This is good, same behaviour for both platforms

File: plugins/native_dio_adapter/.flutter-plugins
Comment (kuhnroyal) [inline plugins/native_dio_adapter/.flutter-plugins:1]: This should not be here

File: plugins/native_dio_adapter/.flutter-plugins-dependencies
Comment (kuhnroyal) [inline plugins/native_dio_adapter/.flutter-plugins-dependencies:1]: This should not be here

File: dio/pubspec.yaml
Comment (kuhnroyal) [inline dio/pubspec.yaml:2]: You can revert the version bump here and in the changelog, we do this automatically when releasing.

File: dio/lib/src/dio_exception.dart
Comment (kuhnroyal) [inline dio/lib/src/dio_exception.dart:161]: Missing trailing comma

File: dio/lib/src/adapters/io_adapter.dart
Comment (kuhnroyal) [inline dio/lib/src/adapters/io_adapter.dart:118]: Can you swap the condition here? That would make it a little bit easier to read.

**Review by kuhnroyal — CHANGES_REQUESTED** [review]
Thanks for your contribution, I left a couple comments.

**Comment by kuhnroyal** [comment]
@cfug/devs In generell I am not sure if this should be considered a breaking change. I know I have code that relies on the current behaviour.

File: dio/pubspec.yaml
Comment (zezo357) [reply dio/pubspec.yaml:2]: So I just set it as unreleased correct?

Comment (kuhnroyal) [reply dio/pubspec.yaml:2]: Yes, in the changelog.

File: plugins/native_dio_adapter/.flutter-plugins
Comment (zezo357) [reply plugins/native_dio_adapter/.flutter-plugins:1]: > This should not be here

to resolve this i think i should revert the commit right?

Comment (kuhnroyal) [reply plugins/native_dio_adapter/.flutter-plugins:1]: Or you can push a new commit where the files get removed, we squash everything in the end.

Comment (zezo357) [reply plugins/native_dio_adapter/.flutter-plugins:1]: > Or you can push a new commit where the files get removed, we squash everything in the end.

done

File: dio/CHANGELOG.md
Comment (kuhnroyal) [inline dio/CHANGELOG.md:8]: This can be removed since you added a change

Comment (zezo357) [reply dio/CHANGELOG.md]: thank you, i forgot it

**Review by kuhnroyal — COMMENTED** [review]
LGTM, just need to figure out if this will break peoples code.

**Comment by zezo357** [comment]
> LGTM, just need to figure out if this will break peoples code.

thank you for the fast and awesome review.

File: dio/CHANGELOG.md
Comment (AlexV525) [inline dio/CHANGELOG.md:8]:
```suggestion
- Fix failing requests throw `DioException`s with `.unknown` instead of `.connectionError` on `SocketException`.
```

File: dio/lib/src/dio_exception.dart
Comment (AlexV525) [inline dio/lib/src/dio_exception.dart:154]: cc @ueman to check if adding the field would affect something.

**Review by AlexV525 — APPROVED** [review]
RSLGTM

File: dio/lib/src/adapters/io_adapter.dart
Comment (ueman) [inline dio/lib/src/adapters/io_adapter.dart:119]: We're now loosing the original stacktrace here. So it should either `Error.throwWithStackTrace()` or the original stacktrace should be passed to the DioException

File: dio/lib/src/dio_exception.dart
Comment (ueman) [reply dio/lib/src/dio_exception.dart:154]: This is safe to do, because it's not a required field

File: dio/lib/src/adapters/io_adapter.dart
Comment (kuhnroyal) [reply dio/lib/src/adapters/io_adapter.dart:119]: Here we are again 😆  I think the original stacktrace doesn't help the user in figuring out where the problem originated and which Dio call caused it. The stacktrace for the `SocketException` always points to some `dart:io`

Comment (AlexV525) [reply dio/lib/src/adapters/io_adapter.dart:119]: Let's just pass it anyway since we'll make a huge refactor using `Error.throwWithStackTrace` in future versions. 🤔

Comment (kuhnroyal) [reply dio/lib/src/adapters/io_adapter.dart:119]: But then we lose the one that is important.

Comment (kuhnroyal) [reply dio/lib/src/adapters/io_adapter.dart:119]: We don't do it for the others (timeout) etc. either.
I created https://github.com/cfug/dio/discussions/1914 to discuss this problem further.

**Comment by zezo357** [comment]
Is something required from me other than the change log text review?

**Comment by AlexV525** [comment]
@kuhnroyal 's review is still requesting changes. PTAL

**Comment by zezo357** [comment]
added the stacktrace


but this fails in the test 

```
                (e) => e.stackTrace
                     .toString()
                    .contains('test/stacktrace_test.dart'),
```
so i commented it, should i remove it? or leave it so it gets tested when a better logic for stacktrace exists.

**Comment by kuhnroyal** [comment]
> @cfug/devs In generell I am not sure if this should be considered a breaking change. I know I have code that relies on the current behaviour.

Any feedback on this?

File: dio/lib/src/adapters/io_adapter.dart
Comment (kuhnroyal) [inline dio/lib/src/adapters/io_adapter.dart:124]: We don't pass the original stacktrace here neither.

**Comment by AlexV525** [comment]
> > @cfug/devs In generell I am not sure if this should be considered a breaking change. I know I have code that relies on the current behaviour.
> 
> Any feedback on this?

I would not consider this as a BC.

File: dio/lib/src/adapters/io_adapter.dart
Comment (kuhnroyal) [reply dio/lib/src/adapters/io_adapter.dart:124]: This was not meant a s a TODO. Please revert, sorry.

File: dio/lib/src/adapters/io_adapter.dart
Comment (kuhnroyal) [inline dio/lib/src/adapters/io_adapter.dart:132]: `StackTrace.current` is wrong, either we take the original stacktrace by catching it. Or none which will lead to a usable default.

Comment (kuhnroyal) [reply dio/lib/src/adapters/io_adapter.dart]: Yea, I understand, sorry.
If the others persist in having the original stacktrace, then using the catched stacktrace is correct.

File: dio/test/stacktrace_test.dart
Comment (kuhnroyal) [inline dio/test/stacktrace_test.dart:1]: @cfug/devs The test changes show that we are changing the stacktrace behavior (again). I suggest to keep the current stacktrace behavior and only change the `DioExceptionType` in the scope of this PR. The rest can be discussed separately.

Comment (zezo357) [reply dio/test/stacktrace_test.dart:1]: just to be sure, so i should remove the stackTrace added in the connection error? correct?

**Comment by zezo357** [comment]
any idea when can this get merged?

**Review by kuhnroyal — APPROVED** [review]
We are not passing the original stacktrace for now, same as before this PR.
Further discussion for this in https://github.com/cfug/dio/discussions/1914

**Comment by kuhnroyal** [comment]
Thanks @zezo357 for your contribution and patience!

**Pull request merged**
